pydm/data_plugins/calc_plugin.py

```diff
@@ -0,0 +1,204 @@
+import functools
+import logging
+import json
+import math
+import numpy as np
+import threading
+
+from pydm.data_plugins.plugin import PyDMPlugin, PyDMConnection
+from qtpy.QtCore import Slot, QThread, Signal, Qt
+from qtpy.QtWidgets import QApplication
+
+import pydm
+
+logger = logging.getLogger(__name__)
+
+
+class CalcThread(QThread):
+    new_data_signal = Signal(dict)
+
+    def __init__(self, config, *args, **kwargs):
+        QThread.__init__(self, *args, **kwargs)
+        self.app = QApplication.instance()
+        self.app.aboutToQuit.connect(self.requestInterruption)
+
+        self.config = config
+
+        self._calculate = threading.Event()
+        self._channels = []
+        self._value = None
+        self._values = None
+        self._connections = None
+        self._expression = self.config.get('expr', '')
+
+        for ch_idx, ch in enumerate(self.config.get('channels', [])):
+            conn_cb = functools.partial(self.callback_conn, ch_idx)
```

Comment:
Like below, I think storing these in a dictionary and accessing by name would make things significantly easier. Most of the time when I code something in Python that requires paying attention to array indices, I end up regretting it - and refactoring it down the line.

Reply:
Addressed via https://github.com/slaclab/pydm/pull/637/commits/c9daee59422ea74d03dfa8ff3c76968a0a6492fa. It turned out way better and leveraging also `defaultdict` allowed me to remove some other lines of code.